walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,181 @@
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long meetingDuration = request.getDuration();
+    Collection<String> requiredAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    Collection<TimeRange> blockedTimes = new HashSet<>();
+    Collection<TimeRange> optionalBlockedTimes = new HashSet<>();
+    Collection<TimeRange> timesToAdd = new HashSet<>();
+
+    // If the meeting is longer than a day, then return empty list
+    if (meetingDuration > TimeRange.WHOLE_DAY.duration()) {
+        List<TimeRange> availableTimes = new ArrayList<TimeRange>(timesToAdd);
+        return availableTimes;
+    }
+    
+    if (!events.isEmpty()) {
+      // For each event, check if anyone attending the request is also attending that event.
+      for (Event event : events) {
+        Set<String> intersection = new HashSet<String>(requiredAttendees);
+        Set<String> optionalIntersection = new HashSet<String>(requiredAttendees);
+        intersection.retainAll(event.getAttendees());
+        optionalIntersection.retainAll(event.getAttendees());
+
+        // If one or more requiredAttendees are attending this event, then add it to blockedTimes
+        // Otherwise, if any optional attendees are attending this event, then add it to optionalBlockedTimes
+        if (!intersection.isEmpty()) {
+          blockedTimes.add(event.getWhen());
+        } else if (!optionalIntersection.isEmpy()) {
+            optionalBlockedTimes.add(event.getWhen());
+        }
+      }
+
+      // Get available times from helper methods
+      timesToAdd = addTimes(timesToAdd, blockedTimes, meetingDuration);
+    
+    } else {
+        // If no one in the required attendees is attending any event in events, then return the whole day
+        timesToAdd.add(TimeRange.WHOLE_DAY);
+    }
+
+    // Return appropriate timeranges, or an empty array if none
+    List<TimeRange> availableTimes = new ArrayList<TimeRange>(timesToAdd);
+    Collections.sort(availableTimes, TimeRange.ORDER_BY_START);
+    return availableTimes;
+  }
+
+  private Collection<TimeRange> addTimes(Collection<TimeRange> timesToAdd, Collection<TimeRange> blockedTimes, long meetingDuration) {
+    
+    // Get a sorted Collection of all "blocked" times with NO OVERLAPS
+    Collection<TimeRange> blockedTimeLine = getBlockedTimeline(blockedTimes);
+    // Create arraylist so we can index into it
+    ArrayList<TimeRange> blockedTimesArrayList = new ArrayList<TimeRange>(blockedTimeLine);
+    // Order the blockedTimeLine by the events that happen first
+    Collections.sort(blockedTimesArrayList, TimeRange.ORDER_BY_START);
+    
+
+    if (blockedTimeLine.isEmpty()) {
+        timesToAdd.add(TimeRange.WHOLE_DAY);
+    } else {
+        // Go through all the blocked times, add non-blocked periods to timesToAdd
+        for (int blockIndex = 0; blockIndex < blockedTimesArrayList.size()+1; blockIndex++) {
```

Comment:
Use <= instead of blockedTimesArrayList.size()+1

Reply:
done! I pushed a new commit that uses <= instead of +1.